crates/runtime/src/model/tool_use.rs

```diff
@@ -641,153 +655,158 @@ fn make_a_stream(
+
+                    // If a tool has finished (i.e. we have all chunks), process them.
+                    if let Some(finish_reason) = &chat_choice.finish_reason {
+                        if matches!(finish_reason, FinishReason::ToolCalls) {
+                            let tool_call_states_clone = Arc::clone(&tool_call_states);
+
+                            let tool_calls_to_process = {
+                                match tool_call_states_clone.lock() {
+                                    Ok(states_lock) => states_lock
+                                        .iter()
+                                        .map(|(_key, tool_call)| tool_call.clone())
+                                        .collect::<Vec<_>>(),
+                                    Err(e) => {
+                                        tracing::error!("Failed to lock tool_call_states: {}", e);
                                         return;
                                     }
                                 }
-                            }
-                            Err(e) => {
-                                if let Err(e) = sender_clone.send(Err(e)).await {
-                                    tracing::error!("Error sending error: {}", e);
+                            };
+
+                            let new_messages = match model
+                                .process_tool_calls_and_run_spice_tools(
+                                    req.messages.clone(),
+                                    tool_calls_to_process,
+                                )
+                                .await
+                            {
+                                Ok(Some(messages)) => messages,
+                                Ok(None) => {
+                                    // No spice tools within returned tools, so return as message in stream.
+                                    finished_choices.push(chat_choice);
+                                    continue;
                                 }
-                                return;
-                            }
-                        };
-                    } else if matches!(finish_reason, FinishReason::Stop)
-                        || matches!(finish_reason, FinishReason::Length)
-                    {
-                        // If complete, return to stream original.
-                        finished_choices.push(chat_choice.clone());
+                                Err(e) => {
+                                    if let Err(e) = sender_clone.send(Err(e)).await {
+                                        tracing::error!("Error sending error: {}", e);
+                                    }
+                                    return;
+                                }
+                            };
+
+                            let mut new_req = req.clone();
+                            new_req.messages.clone_from(&new_messages);
+                            match model.chat_stream(new_req).await {
```

Comment:
Follow on from above, this is an example. Even though it's within an `ai_completion`, I think we would want to refer to this as another `ai_completion` subspan.

Reply:
Similar question above, on L630, with `self.chat_request(new_req).await`.